text/0143-versioning-and-stability.md

```diff
@@ -0,0 +1,128 @@
+Provide maintainers a clear process for developing new, experimental APIs alongside stable APIs. DIfferent packages within the same release will have different levels of stability.
+
+## Relevant architecure
+
+OpenTelemetry is split into signals. Each signal provides a specialized form of observability.
+
+Each signal consists of four types of packages:
+
+**API -** Public interfaces depended upon by 3rd-party libraries and application code instrumented with  OpenTelemetry. Every signal has its own, independent API package. These individual APIs are then bundled up into a Global API for convenience.
+
+**SDK -** The implemention the API. The SDK is only referenced during setup and teardown of an application. SDKs include additional interfaces, such as plugin and lifecycle interfaces, which are not considered part of the API package as they are not imported by shared libraries and application code.
+
+**Semantic Conventions -** A schema defining the attributes which describe common concepts and operations which the signal observes. Note that unlike the API or SDK, stable conventions for all signals may be placed in the same package, as they are often useful across different signals.
+
+**Contrib –** plugins and instrumentation that make use of the API or SDK interfaces, but are not part of the core packages necessary for running OTel. The term contrib specifically refers to the plugins and instrumentation maintained by the OpenTelemetry SIG; it does not refer to third party plugins hosted elsewhere.
+
+## Signal lifecycle
+
+OpenTelemetry is structured around signals. Each signal represents a coherent, stand-alone set of functionality. Each signal follows a lifecycle.
+
+![drawing](img/0143_api_lifecycle.png)
+
+### Lifecycle stages
+
+**Experimental –** Bugs, breaking changes and performance issues may occur.
+
+**Stable –**  Stability guarantees apply, based on component tpye (API, SDK, Conventions, and Contrib). Long term dependencies may now be taken against these packages.
+
+**Deprecated –** this signal has been replaced but is still supported under LTS.
+
+**Removed -** a deprecated signal is no longer supported, and is removed.
+
+All signal components may become stable together, or one by one in the following order: API, SDK, Semantic Conventions, Contrib.
+
+### Stability
+
+Once a signal component is marked as stable, the following rules apply until the end of that signal’s existence.
+
+**API Stability -**
+No backward-incompatible changes to the API are allowed unless the major version number is incremented. All existing API calls must continue to compile and function against all future minor versions of the same major version. ABI compatibility is offered in languages which require it.
+
+**SDK Stability -**
+Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break; ABI compatibility is not required.
```

Comment:
```suggestion
Public portions of the SDK (constructors, configuration, end-user interfaces) must remain backwards compatible. Internal interfaces are allowed to break, i.e., API or ABI compatibility is not required for internal interfaces.
```

This is the same as API right? I don't see any difference in the requirement that public API must be compatible, while internal APIs are allowed to break, with regards to API or SDK. If the intent is that the requirement is softer for SDKs, I don't think we can do that - custom samplers / exporters are good examples of artifacts that will exist outside of OpenTelemetry and that we must not break because of changes to ABI (following the principal of "Ensure that end users stay up to date with the latest release."). Even for just users of management interfaces, we can expect libraries like Spring - we should not create dependency conflicts between versions of Spring and the latest version of OpenTelemetry by having breaking changes.

So I'm not sure I see a reason to split API and SDK stability guarantees into two paragraphs like this, they're the same I think.

Some related conversation at https://github.com/open-telemetry/opentelemetry-java/pull/2212/files#r537932855

Reply:
We're skeptical that we can maintain SDK guarantees as easily as we maintain API guarantees. That's why they are separate.

Spring should never depend on the SDK. Why would that happen?

Reply:
Spring _Sleuth_ (I should have clarified) is a configuration system for observability of Spring apps - users will set some flags in yaml or properties files and spring will configure the SDK on behalf of the user based on those files or what's in the classpath. It has a very similar role to the java autoinstrumentation agent, and does need to depend on the SDK. Unlike the autoinstrumentation agent, it doesn't have the power to bridge the app into an isolated SDK and will depend on our versions.

That being said, it could be possible to allow Sleuth and SDK versions to be tied since they're at the top level - though this needs to be confirmed.

`Sampler` and `SpanExporter` are more important - there definitely will be custom implementations of these. At the Java SIG meeting, we had the thought that we need to identify the SDK APIs that are extension points and we do need to make sure don't cause version conflicts.

Reply:
@anuraaga In the case of Java, I had created https://github.com/open-telemetry/opentelemetry-java/issues/1909 which I think would help with the breakage of anything outside the SDK

Reply:
Totally agree that the SDK needs some stability. We absolutely need to mark out what counts as public and backwards compatible in the SDK.

Reply:
I don't believe that we should enforce complete ABI compatibility for "all SDK" implementations (ie. languages), so I believe that the original comment of "ABI compatibility is not required." is the correct way to phrase this here and any specific language / SDK implementation can then "choose" to add further constraints/support on their provided implementations.

But in general, we should only be guaranteeing that the "public" API's (entry points, configs) are stable as already pointed out in the preceding points with "internal" (implementation level) details (API, Class, functions etc) should always be up for grabs otherwise (depending on the language) you will end up in code bloat city with super fragile implementations.

Reply:
I clarified that ABI compatibility requirements are left to individual languages.